library/core/src/array/mod.rs

```diff
@@ -393,37 +393,81 @@ impl<T, const N: usize> [T; N] {
     where
         F: FnMut(T) -> U,
     {
-        use crate::mem::MaybeUninit;
-        struct Guard<T, const N: usize> {
-            dst: *mut T,
-            initialized: usize,
-        }
-
-        impl<T, const N: usize> Drop for Guard<T, N> {
-            fn drop(&mut self) {
-                debug_assert!(self.initialized <= N);
-
-                let initialized_part =
-                    crate::ptr::slice_from_raw_parts_mut(self.dst, self.initialized);
-                // SAFETY: this raw slice will contain only initialized objects
-                // that's why, it is allowed to drop it.
+        use crate::mem::{align_of, forget, size_of, MaybeUninit, Transmute, TransmuteCopy};
+
+        if align_of::<T>() == align_of::<U>() && size_of::<T>() == size_of::<U>() {
+            union Translated {
+                src: T,
+                dst: U,
+            };
+            struct Guard<const N: usize> {
+                src_dst: *mut Translated,
+                initialized: usize,
+            }
+            impl<T, const N: usize> Drop for Guard<T, N> {
+                fn drop(&mut self) {
+                    debug_assert!(self.initialized <= N);
+
+                    let initialized_part =
+                        crate::ptr::slice_from_raw_parts_mut(self.dst as *mut U, self.initialized);
+                    let todo_part = crate::ptr::slice_from_raw_parts_mut(
+                        self.dst.add(self.initialized) as *mut T,
+                        N - self.initialized,
+                    );
+                    // SAFETY: this raw slice will contain only initialized objects
+                    // that's why, it is allowed to drop it.
+                    unsafe {
+                        crate::ptr::drop_in_place(initialized_part);
+                        crate::ptr::drop_in_place(todo_part);
+                    }
+                }
+            }
+            // reuse array as backing store
+            // SAFETY:
+            let mut src_dst = unsafe { transmute::<_, [MaybeUninit<Translated>; N]>(self) };
+            let mut guard: Guard<U, N> =
+                Guard { src_dst: MaybeUninit::first_ptr_mut(&mut src_dst), initialized: 0 };
+            for i in 0..N {
+                // SAFETY:
```

Comment:
forgot to fill in the comments?

Reply:
Ah yea I'm just ignoring them until I actually run the test because I'm not even sure if transmute is permitted here